%\VignetteIndexEntry{User guide of benchmarkR}
%\VignettePackage{benchmarkR}
%\VignetteEngine{knitr::knitr}
\documentclass[12pt]{article}

<<knitr, echo=FALSE, results="hide", warning=FALSE>>=
library("knitr")
opts_chunk$set(tidy=FALSE,dev="png",fig.show="hide",
               fig.width=4,fig.height=4.5,
               message=FALSE)
@ 

<<style, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
@

<<load_benchmarkR, echo=FALSE>>=
library("benchmarkR")
@
\usepackage{hyperref}
\usepackage{tocloft}
\usepackage{float}
\usepackage{newfloat}
\usepackage[tableposition=top]{caption}
\usepackage{tikz}
\usetikzlibrary{shapes,arrows}

\newcommand{\exitem}[3]{%
  \item \texttt{\textbackslash#1\{#2\}} #3 \csname#1\endcsname{#2}.%
}

\DeclareFloatingEnvironment[name={Example},fileext=lsf,listname={List of Examples}]{example}

\title{User's guide to the ``benchmarkR'' package}
\author{Xiaobei Zhou, Charity Law, Mark D. Robinson}

\begin{document}

\maketitle

\tableofcontents
\listofexamples
\thispagestyle{empty}
\clearpage

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section{Preliminaries}

%\begin{abstract}
% /Bioconductor \cite{Gentleman2004}
\Rpackage{benchmarkR} is an R package designed with a simple purpose: to calculate, evaluate and
visualize the performance of classification method, such as those common in statistical genomics.
For example, much of our research has been related to methods to find differentially expressed genes,
differential splicing, differential methylation and so on.  In principle, \Rpackage{benchmarkR} can
be used for any similar situation, for example, where statistical tests are conducted
across the features (e.g., genes) of a dataset and one wants to contrast sensitivity and false
positive (discovery) rates.
The package would typically be used for synthetic datasets or for datasets that are accompanied by independent
validated ``truth'' (in quotes because validation data are typically measured with some uncertainty).

Researchers already look at the performance of methods through simulations using ROC curves,
precision-recall plots, false detection plots and the like -- so, what does \Rpackage{benchmarkR} offer?
In these plots, we are often interested to know at the same time whether methods are
``calibrated'', that is, do the methods actually achieve the false discovery control that the cutoffs 
used would suggest.  The main idea behind the \Rpackage{benchmarkR} package is to simply augment the
the set of standard plots that many methodologists already look at, with this additional calibration information.  
To do this, we in fact make use of the excellent codebase available in the \Rpackage{ROCR} package for
performance assessment.

This guide provides an overview of the \Rpackage{benchmarkR} package and
detailed information on how to use it across the different types of plots and features to customize the visualizations.

\vspace{1em}
\begin{center}
    \begin{tabular}{ | l | }
      \hline
      \textbf{benchmarkR version:} \Sexpr{packageDescription("benchmarkR")$Version} \\
      \hline
\end{tabular}
\end{center}
\vspace{1em}
\begin{center}
    \begin{tabular}{ | l | }
      \hline
      \textbf{Cittation:} \\
      \\
      XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\\
      XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\\
      XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\\
      \hline
\end{tabular}
\end{center}
\vspace{1em}

% \end{abstract}


\section{A quick start}
In the simplest case, \Rpackage{benchmarkR} would start with a set of P-values and corresponding truth labels, as follows:\\ 
\begin{example}[H]                 
\centering
<<quick, eval=FALSE>>=
library(benchmarkR)
results <- SimResults(pval,labels)
benchmarkR(results)
@
\caption[a quick start]{a quick start} 
\label{quick}
\end{example}

\section{A short summary of \Rpackage{benchmarkR}}
\Rpackage{benchmarkR} is an R package designed to evaluate and visualize classification performance for reference datasets.  Generally speaking, 
the \Rpackage{benchmarkR} package can be divided into two major parts:\\
\begin{enumerate}
  \item \Robject{SimResults}:  a S4 class object to store the P-value, true labels and other necessary values for further analysis and evaluation.
  \item \Rfunction{benchmarkR}: a method to produce a visualized benchmark result including a multi-panel plot of \Robject{rocX}, \Robject{fdX} and \Robject{powerFDR}.
\end{enumerate}
Its whole structure is shown below:\\
  
\begin{center}    
\begin{tikzpicture}[node distance = 2cm, auto]
    % Place nodes
    \node(input) [rectangle, draw] {input data: pval, padj, labels};
    \node(SimResults)[ellipse, draw, below of=input]  {\Robject{SimResults}};
    \node(benchmarkR)[ellipse, draw, below of=SimResults]  {\Robject{benchmarkR}};
    \node(fdX)[diamond, draw, below of=benchmarkR]  {\Robject{fdX}};
    \node(rocX)[diamond, draw, left of=fdX]  {\Robject{rocX}};
    \node(powerFDR)[diamond, draw, right of=fdX]  {\Robject{powerFDR}};
    \node(other)[diamond, draw, right of=powerFDR]  {\Robject{...}};
    % line
    \path [draw, -latex',dashed] (input) -- (SimResults);
    \path [draw, -latex',dashed] (SimResults) -- (benchmarkR);
    \path [draw, -latex',dashed] (benchmarkR) -- (fdX);
    \path [draw, -latex',dashed] (benchmarkR) -- (rocX);
    \path [draw, -latex',dashed] (benchmarkR) -- (powerFDR);
    \path [draw, -latex',dashed] (benchmarkR) -| (other);
\end{tikzpicture}
\end{center}

The individual methods, \Robject{rocX}, \Robject{fdX}, \Robject{powerFDR}, can be called individually and can be highly customized.

%---------------------------------------------------------
\section{Data preparation: build a ``SimResults'' object}
\label{sec:data}

To use the \Rpackage{benchmarkR} package, it is best to build a \Robject{SimResults} object, which is 
just a container for the results. A \Robject{SimResults} object contains three basic input data: 
\begin{itemize}
  \item \Robject{pval}: a vector or matrix containing the P-values
  \item \Robject{padj}: (optionally) a vector or matrix containing the adjusted P-values
  \item \Robject{labels}:  a corresponding vector containing the true class labels (only containing '0' or '1' where '1' is positive and '0' is negative.)
\end{itemize}

The function \Rfunction{SimResults} is just the constructor to build a \Robject{SimResults} object.  To give a practical example, 
we have pre-packaged the simulation results from Zhou et al. (2014) \cite{Zhou2014a} to
highlight the construction of a \Robject{SimResults} object (see also \Rfunction{help("Pickrell")}): \\
%
\begin{example}[H]
\centering
 <<practica,message=TRUE>>=
library(benchmarkR)
data(Pickrell)
re <- SimResults(pval=Pickrell$pval, labels=Pickrell$labels)
re
@
\caption[a practical example -- \Robject{pval} is a matrix]{a practical example -- \Robject{pval} is a matrix}
\label{practical}
\end{example}


In this case, \Robject{padj} is missing. Internally, P-values will be adjusted using a method of choice (default is Benjamini-Hochberg, adjMethod="BH", but others from \Rfunction{p.adjust} can be used). 

%You can also choose a different method for adjusting pvalue, such as:\\
%\begin{example}[H]                 
%\centering
%<<padjMethod,message=TRUE>>=
%re <- SimResults(pval=pval, labels=labels, padjMethod="holm" )
%re
%@
%\caption[set up method for adjusting pvalue when \Robject{padj} is missing]{set up method for adjusting pvalue when \Robject{padj} is missing} 
%\label{padjMethod}
%\end{example}

Note that \Robject{pval} (and \Robject{padj}) can be vectors (for a single method) but also be matrices. 
If matrices are used, each column should represent a score (P-value) a classification model and 
each row should represent a feature. 

%---------------------------------------------------------
\section{Present a benchmark}
\subsection{\Robject{benchmarkR}: a visualized benchmark result}
Following with the last example of Section \ref{sec:data}, a visualized benchmark result can be generated with: \\
\begin{example}[H]                 
\centering
 <<benchmarkR, message=TRUE, fig.show=TRUE, fig.width=10, fig.height=6, dev="pdf">>=
data(Pickrell)
re <- SimResults(pval=Pickrell$pval, labels=Pickrell$labels)
benchmarkR(re)
@
\caption[usage of \Rfunction{benchmarkR}]{usage of \Rfunction{benchmarkR} }
\label{benchmarkR}
\end{example}

The function \Rfunction{benchmarkR} will produce a multi-panel figure of three graphical metrics \Robject{rocX}, \Robject{fdX} and \Robject{powerFDR}. They are defined as:\\
\begin{itemize}
\item \Robject{rocX}:  a simple (partial) ROC curve with given threshold X (FDR, default=0.05)
\item \Robject{fdX}: false discovery number by number of selected outcome positive given a certain of threshold X (FDR, default=0.05)
\item \Robject{powerFDR}: power (TPR) versus achieved FDR given a particular threshold (FDR, default=c(0.001,0.05,0.1))
\end{itemize}

The function \Rfunction{benchmarkR} provides many graphical parameters allowing flexible adjustment of each sub-panel. The graphical parameters of \Rfunction{par()} can directly work in \Rfunction{benchmarkR}, but follows one common rule:\\
\vspace{1em}
\begin{center}
    \begin{tabular}{ | l | }
      \hline
      \textbf{common rule of graphical parameters in \Rfunction{benchmarkR}:} \\
      \\
      a single vector (e.g., \Robject{lwd=1}) that represents all sub-panels in the same way; \\
      a list can represent each sub-panel separately (e.g, \Robject{lwd=list(3,1,3)}).\\
      \hline 
\end{tabular}
\end{center}
\vspace{1em}

An example is shown below: \\ 
 
\begin{example}[H]                 
\centering
 <<parameter, message=TRUE, fig.show=TRUE, fig.width=10, fig.height=6, dev="pdf">>=
benchmarkR(re, lwd=list(3,1,3), cex=2)
@
\caption[graphical parameters in \Rfunction{benchmarkR}]{graphical parameters in \Rfunction{benchmarkR} }
\label{parameter}
\end{example}

In this case ``cex'' are the same ($1$) for all panels, while ``lwd'' are $3$ or $1$.  In addition, ``legend'' is 
plugged in \Rfunction{benchmarkR} function as a graphical parameter that can be also freely customized.    
\begin{example}[H]                 
\centering
 <<legend, message=TRUE, fig.show=TRUE, fig.width=10, fig.height=6, dev="pdf">>=
benchmarkR(re, legend=list(NULL,"topleft","bottomright"))
@
\caption[set up ``legend'' in \Rfunction{benchmarkR}]{set up ``legend'' in \Rfunction{benchmarkR} }
\label{legend}
\end{example}

\subsection{\Rfunction{rocX}: a simple (partial) ROC curve with given threshold}
A simple (partial) ROC plot with threshold X (at the chosen threshold of the method's called FDR point), using the features of \Rpackage{ROCR} is one of the sub-panels of \Rfunction{benchmarkR} \cite{Sing2005}.  The special feature of \Robject{rocX} beyond \Rpackage{ROCR} is that an ``X" point is employed to shows the actualposition along the ROC curve (e.g., FDR=0.05).\\

The constructor function \Rfunction{rocX} will produce an S4 object of \Robject{rocX} that can be simply used as:  
\begin{example}[H]                 
\centering
 <<rocX, message=TRUE, fig.show=TRUE, fig.width=6, fig.height=6, dev="pdf">>=
r <- rocX(re)
@
\caption[\Robject{rocX}]{\Robject{rocX}}
\label{rocX}
\end{example}

\Rfunction{rocX} will call \Robject{prediction} and \Robject{performance} from the \Rpackage{ROCR} package using the P-value
and labels from the \Robject{SimResults} object.  Additionally, the value of TPR and FPR corresponding to the threshold 
X will be calculated from the adjust P-values.  We introduced \Rfunction{rocX} to augment the standard ROC curves in order to show where on the curve a method is operating. \\

For convenience, an \Robject{rocX} object can be re-plotted by function \Rfunction{plot} allowing flexible resetting of graphical parameters from \Rfunction{par()}. \\

There are several special graphical parameters (``cexX'', ``pchX'' and ``colX'') referring to setting of ``X'' point of \Robject{rocX}. You can simply change them as:\\

\begin{example}[H]                 
\centering
 <<replot-rocX, message=TRUE, fig.show=TRUE, fig.width=4.5, fig.height=9, dev="pdf">>=
par(mfrow=c(2,1))
plot(r, lwd=4, cex=4,col=1:3)
plot(r, cexX=3,lwdX=3)
@
\caption[re-plot of \Robject{rocX}, using special graphical parameters]{re-plot of \Robject{rocX}, using special graphical parameters}
\label{replot-rocX}
\end{example}  

In the next examples, additional variations are shown.  For example, you can remove ``X'', if it is not needed.\\

\begin{example}[H]                 
\centering
 <<X, results="hide", fig.show=TRUE, fig.width=9, fig.height=4.5, dev="pdf">>=
par(mfrow=c(1,2))
rocX(re, thresholdX=NULL)
rocX(re, xlim=c(0,.5))
@
\caption[various examples of \Robject{rocX}]{various examples of \Robject{rocX}}
\label{X}
\end{example}   

In addition, \Robject{rocX} allows any flexible transformation formula for input data, in order to have more resolution along the ROC curve.     
\vspace{1em}
\begin{center}
    \begin{tabular}{ | l | }
      \hline
      \textbf{transformation:} \\
      \\
      =``1-x'' (default setting)\\
      =``-log10(x)'' (pval or padj is sparse.) \\
      =``x'' (inverse score) \\
      =``f(x)'' (any other functions) \\   
      \hline 
\end{tabular}
\end{center}
\vspace{1em}


%For instance, you can define your own ``transformation'' function like this way.

%\begin{example}[H]                 
%\centering
% <<transformation, results="hide", fig.show=TRUE, fig.width=6, fig.height=6, dev="pdf">>=
%rocX(re,  transformation="-log10(x^5)")
%@
%\caption[transformation of \Robject{pval} and \Robject{padj}]{transformation of \Robject{pval} and \Robject{padj}}
%\label{transformation}
%\end{example} 

\Robject{rocX} can be separated into several sub-ROC curves by using the ``stratify'' factor, which can be useful to subdivide
results according to other features of the assay (e.g., gene expression strength):
 
 \begin{example}[H]                 
\centering
 <<stratify, results="hide", fig.show=TRUE, fig.width=8, fig.height=8, dev="pdf">>=
f <- as.factor(sample(0:3, nrow(Pickrell$pval), replace=TRUE))
ref <- SimResults(pval=Pickrell$pval, labels=Pickrell$labels,stratify=f)
par(mfrow=c(2,2))
par(mar=c(2,5,3,1),mgp = c(2.6, 1, 0))
rocX(ref)
@
\caption[``stratify'' factor  of \Robject{rocX}]{``stratify'' factor  of \Robject{rocX}}
\label{transformation}
\end{example} 

% res1 <- re[f==0,] 
% res2 <- re[f==1,]
% res3 <- re[f==2,] 
% res4 <- re[f==3,] 
% r1 <- rocX(res1, plot=FALSE)
% r2 <- rocX(res2, plot=FALSE)
% r3 <- rocX(res3, plot=FALSE)
% r4 <- rocX(res4, plot=FALSE) 
% rl <- rocXList(r1=r1, r2=r2, r3=r3, r4=r4)

Multiple plots can be also implemented by building an \Robject{rocXList} as a container for storing a 
collection of multiple \Robject{rocX} objects.
\begin{example}[H]                 
\centering
<<rocXList, results="hide",fig.show="hide", fig.width=6, fig.height=6, dev="pdf">>=
par(mfrow=c(2,2), mar=c(2,5,3,1), mgp=c(2.6, 1, 0))
res <- lapply(0:3, function(u) re[f==u,])
rs <- lapply(res, rocX, plot=FALSE)
rl <- rocXList(rs)
plot(rl)
@
\caption[multiple rocX curves(\Robject{rocXList})]{multiple rocX curves(\Robject{rocXList})}
\label{rocXList}
\end{example} 

The arguments of ``addFun'' and ``addFunLocation'' are introduced into the object 
of \Robject{rocXList} to provide a utility of adding a specified function 
into a specified location for a multi-panel figure. 

\begin{example}[H]                 
\centering
 <<add-fun, results="hide",fig.show=TRUE, fig.width=6, fig.height=6, dev="pdf">>=
par(mfrow=c(2,2))
par(mar=c(2,5,3,1),mgp = c(2.6, 1, 0))
f1 <- 'abline(0,1, col=2)' 
loc1 <- c(1,0,1,0)
f2 <- 'abline(0,1, col=3)'
loca2 <- c(0,1,0,1)
plot(rl, addFun=list(f1,f2), addFunLocation=list(loc1,loca2), 
      main = list("a", "b", "c", "d"), lwd = 1:3, col=1:3)
@
\caption[add function for multiple rocX curves(\Robject{rocXList})]{add function for multiple rocX curves(\Robject{rocXList})}
\label{add-fun}
\end{example} 

``add=TRUE'' is used combining different \Robject{rocX} curves together into one figure.

\begin{example}[H]                 
\centering
 <<add, results="hide",fig.show=TRUE, fig.width=6, fig.height=6, dev="pdf">>=
plot(r1)
plot(r2, lwd=2, add=TRUE)
 @
\caption[combining multiple \Robject{rocX} curves together]{combining multiple \Robject{rocX} curves together}
\label{add}
\end{example} 

\subsection{fdX: false discovery number by number of selected outcome
positive given a threshold}

\Rfunction{fdX} plots cumulative false discoveries versus the number of selected top features (here, using P-value as ranking). The ``X'' point in \Robject{fdX} displays the actual FD number at the method's FDR threshold (e.g., FDR=0.05), allowing users to understand whether a method is adequately controlling its FDR. The constructor function \Rfunction{fdX} will produce an S4 object of \Robject{fdX}. \\

A couple quick examples of \Rfunction{fdX} include:

\begin{example}[H]                 
\centering
<<fdX, message=TRUE, fig.show=TRUE, fig.width=4.5, fig.height=9, dev="pdf">>=
par(mfrow=c(2,1))
f <- fdX(re)
plot(f, lwd=2, col=1:3, pchX=9:11, cex=3, legend=list("topleft"))
@
\caption[\Robject{fdX} examples]{\Robject{fdX} examples}
\label{fdX}
\end{example} 

The \Robject{fdX} function (and objects) use almost all the characteristics of the \Robject{rocX} function, including ``thresholdX'', ``transformation'', special graphical parameters (``cexX'', ``pchX'' and ``colX'')  and so on. A customized example is:\\

\subsection{\Robject{powerFDR}: power (TPR) versus achieved FDR given a threshold}

\Rfunction{powerFDR} calculates and plots power (true positive rate) versus the {\em achieved} FDR using adjusted P-values and the known true positives (labels).  By default, it calculates TPR and achieved FDR at three cutoffs: 0.01, 0.05 and 0.10. \\

The constructor function \Rfunction{powerFDR} will produce an S4 object of \Robject{powerFDR}. A typical usage for an object of \Robject{SimResults} is:\\

\begin{example}[H]                 
\centering 
<<powerFDR, results="hide",message=TRUE, fig.show=TRUE, fig.width=6, fig.height=12, dev="pdf">>=
par(mfrow=c(2,1))
p <- powerFDR(re)
powerFDR(re, threshold=c(0.01,0.02,0.03,0.04,0.05), xlim=c(0,0.2))
@
\caption[\Robject{powerFDR} examples]{\Robject{powerFDR} examples}
\label{powerFDR}
\end{example}

Most of the graphical parameters such as ``col'', ``cex'', ``pch'', etc. from \Robject{par()}, can be directly passed to \Rfunction{plot} for the power-to-achieved-FDR plots. ``pointType'', letter indicating how power-FDR values should be plotted: ``b'' for both points and lines; ``p'' for points only; and ``l" for lines only. ``col.line'' and ``lwd.line'', ``col'' and ``lwd'' of line connecting power-FDR points, if ``pointType'' is either ``b" or ``l". ``lwd.threshold'', ``lty.threshold'' and ``col.threshold'' are ``lwd'', ``lty'' and ``col'' referred to the lines drawn for the threshold.

\begin{example}[H]                 
\centering
<<parameter-powerFDR, results="hide",message=TRUE, fig.show=TRUE, fig.width=6, fig.height=6, dev="pdf">>= 
plot(p, cex=2, pch=c(23,25), col=1:3, main="powerFDR plot", 
     lwd.line=2, cex.axis=1.5, col.threshold=4, lwd.threshold=2)
@          
\caption[graphical parameters of \Robject{powerFDR} plot]{graphical parameters of \Robject{powerFDR} plot}
\label{parameter-powerFDR}
\end{example}

 The FDR is considered to be controlled if the false discovery rate falls below the set threshold; 
if this happens, the power-FDR points are filled-in if plotted. power-FDR points are unfilled, 
or empty, if the FDR is not controlled and is equal to or greater than the cutoff it is assessed at. 
Note that the fill-unfill option is only compatible with 'pch' values of 21 to 25. Visually, this 
means that for a given threshold, a point (of the same color) will be filled if it falls left of 
the threshold line, and will be unfilled if it sits right of the threshold line.\\

The function \Rfunction{powerFDR} can also directly work with input value \Robject{padj}, 
\Robject{padj} and \Robject{labels}. \\

\begin{example}[H]                 
\centering
<<direct-powerFDR, results="hide",message=TRUE, fig.show=TRUE, fig.width=6, fig.height=6, dev="pdf">>= 
col <- c("red", "blue", "green")
# P-values from Method 1
padj <- runif(100)
padj <- sort(padj)
labels <- numeric(100)
labels[1:10] <- 1
powerFDR(padj=padj, labels=labels, threshold=c(0.05, 0.1, 0.5), 
         col=col, plot=TRUE, xlim=c(0,1))
# P-values from Method 2 with larger p-values
padj <- runif(100, min=0.02)
padj <- sort(padj)
powerFDR(padj=padj, labels=labels, threshold=c(0.05, 0.1, 0.5), 
         col=col, add=TRUE, pch=22)
legend("bottomright", pch=c(21, 22), 
       legend=c("Method 1", "Method 2"), pt.bg="black")
@          
\caption[direct usage of \Robject{powerFDR} plot without \Robject{SimResults}]{direct usage of \Robject{powerFDR} plot without \Robject{SimResults}}
\label{direct-powerFDR}
\end{example}



%---------------------------------------------------------
\section{Session info}
%---------------------------------------------------------
<<sessInfo, results="asis", echo=FALSE>>=
toLatex(sessionInfo())
@

\bibliography{myrefs}

\end{document}
